Replace DownloadService with abstract Downloader system

bengelhaupt opened the pull request:
- `Downloader`s implement downloading functionality
- `DownloadItem`s implement download management, making the `DownloadManager` redundant
- `FileDownloader` replaces the former `DownloadService` and is backed by Android's native DownloadManager
- `DownloadAsset` and `DownloadManager` now form `FileDownloadItem`

![Downloading openHPI](https://user-images.githubusercontent.com/26904189/81969445-47711500-961e-11ea-826c-3bd604c320e1.png)

Fixes #107

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/helper/DownloadViewHelper.kt:
```diff
@@ -159,21 +161,13 @@ class DownloadViewHelper(
                     activity.showToast(R.string.toast_no_file_viewer_found)
                 }
-            } else {
+            } ?: run {
```
The block is only executed when `download.openIntent` is `null`. Is it that what you mean?
Another approach would be to hide the "Open" button when the `openIntent` is `null`

tobiasrohloff replied:
I was talking about the usage of the programming syntax:

`if (download.openIntent != null) { ... } else { ... }`
vs
`download.openIntent?.let { intent -> ... } ?: run { ... }`

Usually, I would prefer the first option. The second one unwraps the nullable intent, but that should do the first one automatically as well since the compiler remembers the check, or?

bengelhaupt replied:
Yes you are right

bengelhaupt commented on app/src/main/java/de/xikolo/download/DownloadIdentifier.kt:
```diff
@@ -0,0 +1,6 @@
+package de.xikolo.download
+
+interface DownloadIdentifier {
+
```
It is my personal style. The new autoformatter does not complain. Should I change that?

tobiasrohloff replied:
This seems somehow [undefined or optional](https://developer.android.com/kotlin/style-guide#vertical). Good to know. So I'm fine. Ideally, this would be consistent for the whole project but this would require a custom linter rule.

SanJSp reviewed (approved):
Great work! I really liked the usage of the class diagram in the PR, that cleared up a lot for me 😄

tobiasrohloff reviewed (changes requested):
That's huge, thank's for all the effort!

There is one conflict with master.

bengelhaupt commented on app/src/main/java/de/xikolo/App.kt:
```diff
@@ -97,6 +89,10 @@ class App : Application() {
     }
 
+    private fun configureDownloader() {
+        FileDownloadHandler
```
I found out the initialization takes some time (about a second) when the object is initialized. This happens when the object is accessed the first time. The actual configuration part (just setting the listeners) that takes place in the `init` block does not make much of a difference. I just wanted to make sure the object is already initialized so the UI does not hang. I would leave it this way and add a comment.

tobiasrohloff replied:
Okay, but is it blocking here for a second as well?

bengelhaupt replied:
It's not blocking per se, it's just that for some time, there is no callback.

bengelhaupt commented on app/src/main/java/de/xikolo/download/DownloadIdentifier.kt:
```diff
@@ -0,0 +1,3 @@
+package de.xikolo.download
+
+interface DownloadIdentifier
```
That's because anything can be an identifier for a download. I could have as well used `Any` instead of `DownloadIdentifier` throughout the architecture as the generic type. But having a dedicated interface (even if it's empty) makes it easier to extend the architecture in the future while achieving type safety.

tobiasrohloff replied:
But then you could define an ID with a generic type here that will be defined by subclasses?

bengelhaupt replied:
Yes, but in the architecture, not the ID is the identifier which is passed around, but the `DownloadIdentifier` object.

bengelhaupt replied:
There could also be multiple parts identifying a download

tobiasrohloff replied:
I see, fine.

bengelhaupt commented on app/src/main/java/de/xikolo/download/DownloadStatus.kt:
```diff
@@ -0,0 +1,25 @@
+    enum class State {
+        PENDING, RUNNING, SUCCESSFUL, CANCELLED, FAILED;
+
+        fun and(other: State): State {
```
It defines the state of a download consisting of multiple parts. So e.g. if the subtitle download was successful and the video download failed, the whole download counts as failed. It is tested indirectly via the download item unit tests. I will add a comment.

tobiasrohloff replied:
Ah, good to know.

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadHandler.kt:
```diff
@@ -0,0 +1,240 @@
+                if (download.extras.getString(
+                        REQUEST_EXTRA_SHOW_NOTIFICATION,
+                        true.toString()
+                    ) == true.toString()
```
That's somewhat arkward, because the `Extras` constructor accepts a `Map<String, String>` as arguments and no `putBoolean` method. But it has the method `getBoolean` whicht tries to convert the String value to a Boolean. I now use the `getBoolean` method but a library change then might break something.

tobiasrohloff replied:
Not a blocker, take the most stable variant.

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadHandler.kt:
```diff
@@ -0,0 +1,240 @@
+
+            override fun onProgress(
+                download: Download,
+                etaInMilliSeconds: Long,
```
That is the interface definition and argument naming from the `FetchListener` interface. It is used automatically when auto-generating an interface implementation. Should I change it? The parameter you are talking about is not used.

tobiasrohloff replied:
No, fine. But maybe check what `eta` means in this case and leave a comment.

bengelhaupt replied:
ETA stands for "estimated time of arrival" and probably means when the download can be expected to be finished.

tobiasrohloff replied:
Ah, interesting.

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadHandler.kt:
```diff
@@ -0,0 +1,240 @@
+        if (req.extras.getString(
+                REQUEST_EXTRA_SHOW_NOTIFICATION,
+                true.toString()
+            ) == true.toString()
```
I changed it too.

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/section/VideoPreviewFragment.kt:
```diff
@@ -103,6 +103,11 @@ class VideoPreviewFragment : ViewModelFragment<VideoPreviewViewModel>() {
 
+    override fun onResume() {
+        super.onResume()
+        viewModel.onRefresh()
```
I checked again and this piece of code is not relevant anymore. It solved the problem of re-loading the DownloadViewHelpers when the fragment is resumed but is now not used anymore because they hold a reference to a listener object. Thanks for the catch!

tobiasrohloff replied:
Even better. :)

bengelhaupt commented:
I rebased and incorporated the feedback. the changes are in the [Incorporate Feedback Commit](https://github.com/openHPI/xikolo-android/pull/248/commits/794089f7c94fed773839cb9ac3166ad34d17654e).

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadItem.kt:
```diff
@@ -0,0 +1,300 @@
+    private fun performAction(activity: FragmentActivity, action: () -> Unit): Boolean {
+        return if (storage.isWritable) {
+            if (PermissionManager(activity).requestPermission(
+                    PermissionManager.WRITE_EXTERNAL_STORAGE
```
This is wierd. It looks perfectly fine to me.

tobiasrohloff replied:
At least here I see 4 whitespaces too much.

bengelhaupt replied:
Okay, I re-pushed

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadItem.kt:
```diff
@@ -0,0 +1,301 @@
+        return if (storage.isWritable) {
+            if (
+                PermissionManager(activity).requestPermission(
+                        PermissionManager.WRITE_EXTERNAL_STORAGE
```
Here it told me the opposite: https://github.com/openHPI/xikolo-android/pull/248#discussion_r490266519

bengelhaupt commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadItem.kt:
```diff
@@ -0,0 +1,301 @@
+            if (
+                PermissionManager(activity).requestPermission(
+                        PermissionManager.WRITE_EXTERNAL_STORAGE
+                    ) == 1
```
Opposite: https://github.com/openHPI/xikolo-android/pull/248#discussion_r490266527

tobiasrohloff commented on app/src/main/java/de/xikolo/download/filedownload/FileDownloadItem.kt:
```diff
@@ -0,0 +1,301 @@
+            if (
+                PermissionManager(activity).requestPermission(
+                    PermissionManager.WRITE_EXTERNAL_STORAGE
+                ) == 1
```
Here it looks correct.

bengelhaupt replied:
I pushed again with this format
